pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -691,47 +707,58 @@ func (p *PriorityQueue) podsCompareBackoffCompleted(p1, p2 interface{}) bool {
 func (p *PriorityQueue) NumUnschedulablePods() int {
 	p.lock.RLock()
 	defer p.lock.RUnlock()
-	return len(p.unschedulableQ.pods)
+	return len(p.unschedulableQ.unschedulablePods)
 }
 
 // UnschedulablePodsMap holds pods that cannot be scheduled. This data structure
 // is used to implement unschedulableQ.
 type UnschedulablePodsMap struct {
-	// pods is a map key by a pod's full-name and the value is a pointer to the pod.
-	pods    map[string]*v1.Pod
-	keyFunc func(*v1.Pod) string
+	// pods is a map key by a pod's full-name and the value is a pointer to the unschedulablePod.
+	unschedulablePods map[string]*unschedulablePod
+	keyFunc           func(*v1.Pod) string
+	clock             util.Clock
+}
+
+// unschedulablePod contains pointer to the pod and the last update time in unschedulableQ.
+type unschedulablePod struct {
```

Comment:
We don't need this type and don't need to record `lastUpdate` time. Last scheduling attempt is already recorded in the pod status. Please look at `podTimestamp` to find how you can get the last scheduling attempt. In fact, you should be able to use the same function.
Without this type, `UnschedulablePodsMap` can remain the same as it was before.

Reply:
At first I tried to use podTimestamp as the last scheduling attempt, But I found that podTimestamp would not change after another scheduling attempt. 
```go
func (p *podConditionUpdater) Update(pod *v1.Pod, condition *v1.PodCondition) error {
	klog.V(3).Infof("Updating pod condition for %s/%s to (%s==%s, Reason=%s)", pod.Namespace, pod.Name, condition.Type, condition.Status, condition.Reason)
	if podutil.UpdatePodCondition(&pod.Status, condition) {
		_, err := p.Client.CoreV1().Pods(pod.Namespace).UpdateStatus(pod)
		return err
	}
	return nil
}
```
The LastTransitionTime is updated only when the condition changes. It will not be updated after scheduling attempt failed again Because the condition hasn't changed. (PodScheduled)
```go
// UpdatePodCondition updates existing pod condition or creates a new one. Sets LastTransitionTime to now if the
// status has changed.
// Returns true if pod condition has changed or has been added.
func UpdatePodCondition(status *v1.PodStatus, condition *v1.PodCondition) bool {
	condition.LastTransitionTime = metav1.Now()
	// Try to find this pod condition.
	conditionIndex, oldCondition := GetPodCondition(status, condition.Type)

	if oldCondition == nil {
		// We are adding new pod condition.
		status.Conditions = append(status.Conditions, *condition)
		return true
	}
	// We are updating an existing condition, so we need to check if it has changed.
	if condition.Status == oldCondition.Status {
		condition.LastTransitionTime = oldCondition.LastTransitionTime
	}

	isEqual := condition.Status == oldCondition.Status &&
		condition.Reason == oldCondition.Reason &&
		condition.Message == oldCondition.Message &&
		condition.LastProbeTime.Equal(&oldCondition.LastProbeTime) &&
		condition.LastTransitionTime.Equal(&oldCondition.LastTransitionTime)

	status.Conditions[conditionIndex] = *condition
	// Return true if one of the fields have changed.
	return !isEqual
}
```
The above is what I saw. I don't know if there are any mistakes in understanding. If there are any mistakes, I hope you can correct them. Thank you very much. @bsalamat

Reply:
You are right. https://github.com/kubernetes/kubernetes/pull/72619 is sent recently to change lastProbeTime every time a pod is retried. I think we can use that after the PR is merged.